### PR TITLE
Correct data assignment with multiple list indices 

### LINKED ISSUE #217: Incorrect data assignment with some multiple list indices
When multiple list indices are provided to a data assignment (i.e. `Data.__setitem__`), the assignment works if all of the lists are strictly monotonically increasing, but only for some cases cases where they are not:

```python
>>> import cfdm
>>> d = cfdm.example_field(0).data
>>> d[[0, 3, 4], [1, 7, 6]].array 
array([[0.034, 0.029, 0.024],
       [0.059, 0.017, 0.009],
       [0.036, 0.013, 0.034]])
>>> d[[0, 3, 4], [1, 7, 6]] = -1       # all lists strictly monotonically increasing => OK
>>> d[[0, 3, 4], [1, 7, 6]].array 
array([[-1., -1., -1.],
       [-1., -1., -1.],
       [-1., -1., -1.]])
```

```python
>>> import cfdm
>>> d = cfdm.example_field(0).data
>>> d[[0, 3, 4], [1, 6, 7]].array 
array([[0.034, 0.029, 0.024],
       [0.059, 0.017, 0.009],
       [0.036, 0.013, 0.034]])
>>> d[[0, 3, 4], [1, 7, 6]] = -1 
>>> d[[0, 3, 4], [1, 7, 6]].array        # one list not strictly monotonically increasing => OK (edited from "not OK"!)
array([[-1., -1., -1.],
       [-1., -1., -1.],
       [-1., -1., -1.]])
```

```python
>>> d = cfdm.example_field(0).data
>>> d[[4, 3, 0], [7, 6, 1]] = -1
>>> d[[4, 3, 0], [7, 6, 1]].array       # both lists strictly monotonically decreasing => NOT OK
array([[ 0.013,  0.034,  0.036],
       [ 0.017,  0.009,  0.059],
       [ 0.029,  0.024, -1.   ]])
```
```python
>>> d = cfdm.example_field(0).data
>>> d[[4, 3, 0], [7, 6, 1]] = -1
>>> d[[4, 3, 0], [7, 6, 1]].array     # NOT OK
array([[ 0.013,  0.034,  0.036],
       [ 0.017,  0.009,  0.059],
       [ 0.029,  0.024, -1.   ]])
```
```python
>>> d = cfdm.example_field(0).data
>>> d[[0, 4, 3], [1, 7, 6]] = -1
>>> d[[0, 4, 3], [1, 7, 6]].array   # OK, though!
array([[-1., -1., -1.],
       [-1., -1., -1.],
       [-1., -1., -1.]])
```

PR to follow (after cfdm has been transferred to `main`) (edit: https://github.com/davidhassell/cfdm/tree/data-setitem-multiple-list-indices-2, lest I forget)

```python
>>> cfdm.environment(paths=False)
Platform: Linux-5.14.0-1052-oem-x86_64-with-glibc2.31
HDF5 library: 1.12.1
netcdf library: 4.8.1
Python: 3.9.12
netCDF4: 1.6.0
numpy: 1.22.3
cfdm.core: 1.10.0.0
cftime: 1.6.1
netcdf_flattener: 1.2.0
cfdm: 1.10.0.0
```

### PR DESCRIPTION
**Pull request by davidhassell**
Fixes #217

### CODE REVIEW

**Review by sadielbartholomew — APPROVED** [review]
Fixes the underlying issue and adds new testing with good coverage to capture any potential issues along those lines in future. I've raised some minor comments and questions, but overall this is all good. Please merge when ready, thanks!

File: cfdm/data/data.py [hunk 1]
```diff
@@ -858,64 +858,198 @@ def _set_CompressedArray(self, array, copy=True):
...
-            # --------------------------------------------------------
             # At least two axes have list-of-integers indices so we
-            # can't do a normal numpy assignment
-            # --------------------------------------------------------
+            # can't do a normal assignment.
+            #
+            # The brute-forace approach would be to do a seperate
+            # assignment to each set of elemants of 'array' that is
+            # defined by a every possible combination of the integers
+            # defined by the two of index lists.
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cfdm/data/data.py:951]: ... and I thank you for it!

File: cfdm/data/data.py [hunk 2]
```diff
@@ -858,64 +858,198 @@ def _set_CompressedArray(self, array, copy=True):
...
+            # This can be reduced by a factor of ~2 per axis that has
+            # list indices if we convert it to a sequence of "size 2"
+            # slices (with a "size 1" slice at the end if there are an
+            # odd number of list elements).
+            #
+            # In the above example, the input list index [1, 2, 4, 5]
+            # can be mapped to two slices: slice(1,3,1), slice(4,6,1);
+            # the input list index [8, 9] is mapped to slice(8,10,1)
+            # and only 2 seperate assignments of 40 elements each are
+            # needed.
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cfdm/data/data.py:966]: `slice(0, 10)` is not a list index, so doesn't need to be altered. The "40" in "2 seperate assignments of 40 elements each" is derived from 2x2x10, where the "10" is coming from the non-list index `slice(0, 10)`.

File: cfdm/data/data.py [hunk 3]
```diff
@@ -858,64 +858,198 @@ def _set_CompressedArray(self, array, copy=True):
...
+        >>> Data._set_subspace(a, [[1, 4 ,3], [7, 6, 1]],
+        ...                    np.array([[-1, -2, -3]]))
+        >>> print(a)
+        [[ 0  1  2  3  4  5  6  7]
+         [ 8 -3 10 11 12 13 -2 -1]
+         [16 17 18 19 20 21 22 23]
+         [24 -3 26 27 28 29 -2 -1]
+         [32 -3 34 35 36 37 -2 -1]]
+
+        >>> a = np.arange(40).reshape(5, 8)
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cfdm/data/data.py:902]: https://github.com/NCAS-CMS/cfdm/pull/220/commits/00ce7d509e2081ab0a252f0f88da3a8ac733c048

File: cfdm/data/data.py [hunk 4]
```diff
@@ -858,64 +858,198 @@ def _set_CompressedArray(self, array, copy=True):
...
     @classmethod
-    def _set_subspace(cls, array, indices, value):
-        """Set a subspace of the data array defined by indices."""
+    def _set_subspace(cls, array, indices, value, orthogonal_indexing=True):
+        """Assign to a subspace of an array.
+
+        :Parameters:
+
+            array: array_like
+                The array to be assigned to. Must support fancy
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cfdm/data/data.py]: OK, best not to use that phrase then, perhaps. How about: https://github.com/NCAS-CMS/cfdm/pull/220/commits/72f2071fe313f2faefd27124815dcd31252058b6

File: cfdm/data/data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cfdm/data/data.py:970]: Oops, that's  a very good point that I did indeed miss :smile: Thanks for clarifying.

File: cfdm/data/data.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cfdm/data/data.py]: Yes, that's a great replacement! Nice idea. Marking as resolved, though there is an instance of this in the description for `value: array_like` too that you might wish to convert.

File: cfdm/data/data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cfdm/data/data.py:902]: Perfect, thanks. Marking as resolved.

**Review by sadielbartholomew — APPROVED** [review]
All feedback raised has been addressed and I've done a quick sanity check on the final overall state of the PR. Good to merge now, thanks!

**Pull request merged**